sdk/core/azure-core/src/cryptography/md5.cpp

```diff
@@ -16,156 +16,187 @@
+    ContextSize = objectLength;
+    DWORD hashLength = 0;
+    if (!BCRYPT_SUCCESS(
+            m_status = BCryptGetProperty(
+                Handle,
+                BCRYPT_HASH_LENGTH,
+                reinterpret_cast<PBYTE>(&hashLength),
+                sizeof(hashLength),
+                &dataLength,
+                0)))
     {
-      throw std::runtime_error("BCryptCreateHash failed");
+      throw std::runtime_error("BCryptGetProperty failed with code" + m_status);
     }
+    HashLength = hashLength;
   }
 
-  Md5Hash::~Md5Hash()
+  ~Md5AlgorithmProvider()
   {
-    _detail::Md5HashContext* md5Context = static_cast<_detail::Md5HashContext*>(m_md5Context);
-    BCryptDestroyHash(md5Context->hashHandle);
-    delete md5Context;
+    if (Handle)
+    {
+      BCryptCloseAlgorithmProvider(Handle, 0);
+    }
   }
-
-  void Md5Hash::OnAppend(const uint8_t* data, std::size_t length)
+};
+
+Md5AlgorithmProvider const& GetMD5AlgorithmProvider()
+{
+  static Md5AlgorithmProvider instance;
+  return instance;
+}
+
+class Md5BCrypt : public Azure::Core::Cryptography::Hash {
+private:
+  NTSTATUS m_status = 0;
+  BCRYPT_HASH_HANDLE m_hashHandle = nullptr;
+  std::size_t m_hashLength = 0;
+  std::string m_buffer;
+
+  void OnAppend(const uint8_t* data, std::size_t length)
   {
-    _detail::Md5HashContext* md5Context = static_cast<_detail::Md5HashContext*>(m_md5Context);
-
-    NTSTATUS status = BCryptHashData(
-        md5Context->hashHandle,
-        reinterpret_cast<PBYTE>(const_cast<uint8_t*>(data)),
-        static_cast<ULONG>(length),
-        0);
-    if (!BCRYPT_SUCCESS(status))
+    if (!BCRYPT_SUCCESS(
+            m_status = BCryptHashData(
+                m_hashHandle,
+                reinterpret_cast<PBYTE>(const_cast<uint8_t*>(data)),
+                static_cast<ULONG>(length),
+                0)))
     {
-      throw std::runtime_error("BCryptHashData failed");
+      throw std::runtime_error("BCryptHashData failed with code" + m_status);
     }
   }
 
-  std::vector<uint8_t> Md5Hash::OnFinal(const uint8_t* data, std::size_t length)
+  std::vector<uint8_t> OnFinal(const uint8_t* data, std::size_t length)
   {
     OnAppend(data, length);
-    _detail::Md5HashContext* md5Context = static_cast<_detail::Md5HashContext*>(m_md5Context);
+
     std::vector<uint8_t> hash;
-    hash.resize(md5Context->hashLength);
-    NTSTATUS status = BCryptFinishHash(
-        md5Context->hashHandle,
-        reinterpret_cast<PUCHAR>(&hash[0]),
-        static_cast<ULONG>(hash.size()),
-        0);
-    if (!BCRYPT_SUCCESS(status))
+    hash.resize(m_hashLength);
+    if (!BCRYPT_SUCCESS(
+            m_status = BCryptFinishHash(
+                m_hashHandle,
+                reinterpret_cast<PUCHAR>(&hash[0]),
+                static_cast<ULONG>(hash.size()),
+                0)))
     {
-      throw std::runtime_error("BCryptFinishHash failed");
+      throw std::runtime_error("BCryptFinishHash failed with code" + m_status);
```

Comment:
needs some spacing or punctuation so it doesn't read like: `BCryptFinishHash failed with code4`

Same issue everywhere else.

```suggestion
      throw std::runtime_error("BCryptFinishHash failed with code: " + m_status + ".");
```